velox/expression/CastExpr.cpp

```diff
@@ -123,6 +123,44 @@ void applyIntToDecimalCastKernel(
   });
 }
 
+#define VELOX_DYNAMIC_DECIMAL_TYPE_DISPATCH(       \
+    TEMPLATE_FUNC, decimalTypePtr, ...)            \
+  [&]() {                                          \
+    if (decimalTypePtr->isLongDecimal()) {         \
+      return TEMPLATE_FUNC<int128_t>(__VA_ARGS__); \
+    } else {                                       \
+      return TEMPLATE_FUNC<int64_t>(__VA_ARGS__);  \
+    }                                              \
+  }()
+
+template <typename FromNativeType, TypeKind ToKind>
+VectorPtr applyDecimalToIntegralCast(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& fromType,
+    const TypePtr& toType) {
+  using To = typename TypeTraits<ToKind>::NativeType;
+
+  VectorPtr result;
+  context.ensureWritable(rows, toType, result);
+  (*result).clearNulls(rows);
+  auto resultBuffer = result->asUnchecked<FlatVector<To>>()->mutableRawValues();
+  const auto precisionScale = getDecimalPrecisionScale(*fromType);
+  const auto simpleInput = input.as<SimpleVector<FromNativeType>>();
+  const auto scaleFactor = DecimalUtil::kPowersOfTen[precisionScale.second];
+  context.applyToSelectedNoThrow(rows, [&](int row) {
+    int128_t integralPart = simpleInput->valueAt(row) / scaleFactor;
+    if (integralPart > std::numeric_limits<To>::max() ||
+        integralPart < std::numeric_limits<To>::min())
+      VELOX_FAIL(makeErrorMessage(input, row, toType));
```

Comment:
This needs to be a USER error.

Would you specify that value is out of bounds in the error message to help debugging?

Reply:
Done.